src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -318,102 +326,132 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
             }
         }
 
-        private int FindEntry(TKey key)
+        private unsafe ref TValue FindValue(TKey key)
         {
             if (key == null)
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
             }
 
-            int i = -1;
             int[]? buckets = _buckets;
-            Entry[]? entries = _entries;
-            int collisionCount = 0;
+            ref Entry entry = ref Unsafe.AsRef<Entry>(null);
             if (buckets != null)
             {
-                Debug.Assert(entries != null, "expected entries to be != null");
+                Debug.Assert(_entries != null, "expected entries to be != null");
                 IEqualityComparer<TKey>? comparer = _comparer;
                 if (comparer == null)
                 {
                     uint hashCode = (uint)key.GetHashCode();
-                    // Value in _buckets is 1-based
-                    i = buckets[hashCode % (uint)buckets.Length] - 1;
+                    int i = buckets[hashCode % (uint)buckets.Length];
                     if (default(TKey)! != null) // TODO-NULLABLE: default(T) == null warning (https://github.com/dotnet/roslyn/issues/34757)
                     {
                         // ValueType: Devirtualize with EqualityComparer<TValue>.Default intrinsic
-                        while (true)
+                        Entry[]? entries = _entries;
+                        int collisionCount = 0;
```

Comment:
This duplication of some of the locals looks very unnatural. Is it compensating for a jit bug?

Reply:
Register pressure; I could move them up one level (above `default(TKey)! != null`) but after `buckets[hashCode % (uint)buckets.Length]`

At this point `buckets` and `comparer` have been discarded; where they were originally assigned their lifetime overlapped and caused a stack spill.